src/ADAL.PCL/AuthenticationContext.cs

```diff
@@ -194,428 +216,789 @@ public Guid CorrelationId
-                ExtendedLifeTimeEnabled = this.ExtendedLifeTimeEnabled,
+                ExtendedLifeTimeEnabled = ExtendedLifeTimeEnabled,
+                RequestId = Telemetry.GetInstance().CreateRequestId()
             };
-            var handler = new AcquireTokenInteractiveHandler(requestData, redirectUri, parameters, userId, extraQueryParameters, this.CreateWebAuthenticationDialog(parameters));
-            return await handler.RunAsync().ConfigureAwait(false);
+
+            Telemetry.GetInstance().StartEvent(requestData.RequestId, "api_event");
+
+            AcquireTokenInteractiveHandler handler = new AcquireTokenInteractiveHandler(requestData, redirectUri,
+                parameters, userId,
+                extraQueryParameters, CreateWebAuthenticationDialog(parameters));
+            AuthenticationResult result = null;
+
+            try
+            {
+                result = await handler.RunAsync().ConfigureAwait(false);
+            }
+            finally
+            {
+                ApiEventHelper(result, requestData, ApiId, extraQueryParameters);
+            }
+
+            return result;
         }
 
-        private async Task<AuthenticationResult> AcquireTokenSilentCommonAsync(string resource, ClientKey clientKey, UserIdentifier userId, IPlatformParameters parameters)
+        private async Task<AuthenticationResult> AcquireTokenSilentCommonAsync(string resource,
+            ClientKey clientKey,
+            UserIdentifier userId,
+            IPlatformParameters parameters,
+            string ApiId)
         {
             RequestData requestData = new RequestData
             {
                 Authenticator = Authenticator,
-                TokenCache = this.TokenCache,
+                TokenCache = TokenCache,
                 Resource = resource,
-                ExtendedLifeTimeEnabled = this.ExtendedLifeTimeEnabled,
-                ClientKey = clientKey
+                ExtendedLifeTimeEnabled = ExtendedLifeTimeEnabled,
+                ClientKey = clientKey,
+                RequestId = Telemetry.GetInstance().CreateRequestId()
             };
 
-            var handler = new AcquireTokenSilentHandler(requestData, userId, parameters);
-            return await handler.RunAsync().ConfigureAwait(false);
+            Telemetry.GetInstance().StartEvent(requestData.RequestId, "api_event");
+
+            AcquireTokenSilentHandler handler = new AcquireTokenSilentHandler(requestData, userId, parameters);
+            AuthenticationResult result = null;
+
+            try
+            {
+                result = await handler.RunAsync().ConfigureAwait(false);
+            }
+            finally
+            {
+                ApiEventHelper(result, requestData, ApiId, null);
+            }
+
+            return result;
+        }
+
+        private void ApiEventHelper(AuthenticationResult result, RequestData requestData, string ApiId,string extraQueryParameters)
```

Comment:
>ApiEventHelper [](start = 21, length = 14)

Consider a name that explains what the method is doing. Calling it a helper is too generic.